File: src/main/resources/static/python/ascending/BubbleSortAscending.py
import random
import time
import csv
import os

def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

def write_to_file(filename, arr):
    with open(filename, 'w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(arr)

def read_from_file(filename):
    if not os.path.exists(filename):
        print(f'File not found: {filename}')
        return []
    with open(filename, 'r') as file:
        reader = csv.reader(file)
        return list(map(int, next(reader)))

def main():
    # Get the array size dynamically from user input
    try:
        n = int(input(""))  # Get array size from user input
        if n <= 0:
            raise ValueError
    except ValueError:
        print("Array size must be a positive integer.")
        os.sys.exit(1)

    # Navigate two levels up to reach the static directory
    base_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../'))
    original_file = os.path.join(base_path, 'original_array.csv')
    sorted_file = os.path.join(base_path, 'sorted_result.csv')

    # Always create a new array based on user input
    arr = [random.randint(1, 100) for _ in range(n)]
    write_to_file(original_file, arr)

    start_time = time.time()

    bubble_sort(arr)

    write_to_file(sorted_file, arr)

    end_time = time.time()
    total_time = end_time - start_time
    
    # # Output results with HTML-style links
    # print(f'Original array file: <a href="/static/original_array.csv">original_array.csv</a>')
    # print(f'Sorted array file: <a href="/static/sorted_result.csv">sorted_result.csv</a>')
    print(f'Time taken: {total_time:.3f} seconds.')

if __name__ == "__main__":
    main()

File: src/main/resources/static/cpp/ascending/BubbleSortAscending.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <ctime>
#include <cstdlib>
#include <string>
#include <filesystem>

// Function to perform Bubble Sort
void bubbleSort(std::vector<int>& arr) {
    for (size_t i = 0; i < arr.size() - 1; i++) {
        for (size_t j = 0; j < arr.size() - i - 1; j++) {
            if (arr[j + 1] < arr[j]) {
                std::swap(arr[j], arr[j + 1]);
            }
        }
    }
}

// Function to write an array to a CSV file
void writeToFile(const std::string& filename, const std::vector<int>& arr) {
    std::ofstream file(filename);
    for (size_t i = 0; i < arr.size(); i++) {
        file << arr[i];
        if (i < arr.size() - 1) {
            file << ",";
        }
    }
    file << std::endl;
    file.close();
}

// Function to read an array from a CSV file
std::vector<int> readFromFile(const std::string& filename) {
    std::vector<int> arr;
    std::ifstream file(filename);
    std::string line;
    if (std::getline(file, line)) {
        size_t start = 0;
        size_t end = line.find(',');
        while (end != std::string::npos) {
            arr.push_back(std::stoi(line.substr(start, end - start)));
            start = end + 1;
            end = line.find(',', start);
        }
        arr.push_back(std::stoi(line.substr(start, end)));
    }
    return arr;
}

int main() {
    int n;
    // std::cout << "Enter the size of the array: ";
    std::cin >> n;

    // Get the current path and navigate to the static directory
    std::filesystem::path currentPath = std::filesystem::current_path();
    std::filesystem::path basePath = currentPath.parent_path().parent_path().parent_path() / "static";

    std::string originalFile = (basePath / "original_array.csv").string();
    std::string sortedFile = (basePath / "sorted_result.csv").string();

    std::vector<int> arr;

    // Always generate a new array based on user input
    arr.resize(n);
    std::srand(std::time(nullptr));
    for (int i = 0; i < n; i++) {
        arr[i] = std::rand() % 100 + 1;
    }
    writeToFile(originalFile, arr);

    clock_t startTime = clock();

    // Sort the array
    bubbleSort(arr);

    // Write the sorted array to file
    writeToFile(sortedFile, arr);

    clock_t endTime = clock();
    double totalTime = static_cast<double>(endTime - startTime) / CLOCKS_PER_SEC;

    // // Output results
    // std::cout << "Original array file: " << originalFile << std::endl;
    // std::cout << "Sorted array file: " << sortedFile << std::endl;
    std::cout << "Time taken: " << totalTime << " seconds." << std::endl;

    return 0;
}
